plugins/channel-safely/channel-manager.cpp

```diff
@@ -39,13 +54,87 @@ void ChannelManager::manage_group(const Group &group, bool set_marker_mode, bool
             marker_mode = false;
         }
     }
-    for (auto &designation: group) {
-        manage_one(group, designation, true, marker_mode);
+    // cavein prevention
+    bool cavein_possible = false;
+    //uint8_t least_access = 10;
+    std::unordered_map<df::coord, uint8_t> cavein_candidates;
+    if (!marker_mode) {
+        /* To prevent cave-ins we're looking at accessibility of tiles with open space below them
+         * If it has space below, it has somewhere to fall
+         * Accessibility tells us how close to a cave-in a tile is, low values are at risk of cave-ins
+         * To count access, we find a random miner dwarf and count how many tile neighbours they can path to
+         * */
+        df::unit* miner = nullptr;
+        auto size = df::global::world->units.active.size();
+        std::vector<std::size_t> indices(size);
+        for (std::size_t i = 0; i < size; ++i) {
+            indices[i] = i;
+        }
+        // Shuffle the indices in random order
+        std::random_device rd;
+        std::mt19937 gen(rd());
+        std::shuffle(indices.begin(), indices.end(), gen);
+        for (auto idx : indices) {
+            auto unit = df::global::world->units.active[idx];
+            if (unit->status.labors[df::unit_labor::MINE]) {
+                // we found a random miner, time to move on
+                miner = unit;
+                break;
+            }
+        }
+        // Analyze designations
+        for (const auto &pos: group) {
+            df::coord below(pos);
+            below.z--;
+            auto ttype = *Maps::getTileType(below);
+            // we can skip designations already queued for insta-digging
+            if (CSP::dignow_queue.count(pos)) continue;
+            if (DFHack::isOpenTerrain(ttype) || DFHack::isFloorTerrain(ttype)) {
+                cavein_possible = config.prevent_caveins;
+                auto access = count_accessibility(miner->pos, pos);
+                if (access == 0) {
+                    if (config.insta_dig){
+                        CSP::dignow_queue.emplace(pos);
+                    } else {
+                        // todo: engage dig management, swap channel designations for dig
```

Comment:
is this todo in this PR or later?

Reply:
likely later, depends if a solution is found for the disjointed play problem I brought up on the issue